### PR TITLE
Helper functions to access additionalProblems

### PR DESCRIPTION
**Pull request by maximiliankaul**

### CODE REVIEW

**Comment by maximiliankaul** [comment]
Should we rename the functions so that `problems` returns everything? Would make more sense to me. And then hide the other two as internal.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt [hunk 1]
```diff
@@ -612,6 +612,27 @@ val Node?.assigns: List<AssignExpression>
 val Node?.problems: List<ProblemNode>
     get() = this.allChildren()
 
+/**
+ * Returns all [Node] children in this graph, where the [Node.additionalProblems] field is not
+ * empty.
+ */
+val Node?.additionalProblemsNotEmpty: List<Node>
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt:619]: nodesWithProblems?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt [hunk 2]
```diff
@@ -612,6 +612,27 @@ val Node?.assigns: List<AssignExpression>
...
+    get() = this.allChildren<Node> { node -> node.additionalProblems.isNotEmpty() }
+
+/**
+ * Returns all [ProblemNode] children - stored in a [Node.additionalProblems] field - in this graph,
+ * starting with this [Node].
+ */
+val Node?.additionalProblemsUnpacked: List<ProblemNode>
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt:626]: allProblems?

**Review by oxisto — COMMENTED** [review]
The naming is slightly confusing. Some tests would be good to better see the expected outcome

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt (hunk 2 above)
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt:626]: ah. its only “all additional problems”

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt (hunk 1 above)
Comment (maximiliankaul) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt]: Please check the naming again. I'm quite happy with the current name for the additional problem nodes. However, the name for the problem nodes directly in the graph is a bit misleading. Suggestions?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt (hunk 2 above)
Comment (maximiliankaul) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt]: Yes. And I made it "private", as I don't see an external use case.

**Comment by maximiliankaul** [comment]
I also have a version of this with only one instead of the two graph traversals. It is about 5% faster overall (the entire `analyze` function + `.problems` call). However, it is less readable...

```kotlin
val Node?.fastProblems: List<ProblemNode>
    get() {
        val relevantNodes =
            this.allChildren<Node> { it is ProblemNode || it.additionalProblems.isNotEmpty() }

        val result = mutableListOf<ProblemNode>()

        relevantNodes.forEach {
            if (it.additionalProblems.isNotEmpty()) {
                result += it.additionalProblems
            }
            if (it is ProblemNode) {
                result += it
            }
        }

        return result
    }
```

**Review by oxisto — APPROVED** [review]

**Pull request merged**